Changed spell-checking to be plugin-based.

dmoonfire opened the pull request:
* Changed the package to allow for external packages to provide additional checking. (Closes #74)
      - Updated the Task-based system to handle passing known words to the background task.
      - Two default plugins are included: system-based dictionaries and "known words".
      - Suggestions and "add to dictionary" are also provided via interfaces. (Closes #10)
      - Modified various calls so they are aware of the where the buffer is located.
    * Modified system to allow for multiple plugins/checkers to identify correctness.
      - Incorrect words must be incorrect for all checkers.
      - If no checker has an opinion, it is considered correct.
      - Any checker that treats a word as valid is considered valid for the buffer.
    * Extracted system-based dictionary support into separate checker.
      - System dictionaries can now check across multiple system locales.
      - Locale selection can be changed via package settings. (Closes #21)
      - Multiple locales can be selected. (Closes #11)
      - External search paths can be used for Linux and OS X.
      - Default language is based on the process environment, with a fallback to the browser, before finally using `en-US` as a fallback.
    * Extracted hard-coded approved list into a separate checker.
      - User can add additional "known words" via settings.
      - Added an option to add more known words via the suggestion dialog.
    * Updated ignore files and added EditorConfig settings for development.
    * Various coffee-centric formatting.

dmoonfire commented:
Commit https://github.com/atom/spell-check/commit/833330f48acff00a605f7efad502f2ddcc23a6e5 includes a slightly reworked default language selection based on conversations from [this discussion thread](https://discuss.atom.io/t/need-help-with-testing-spell-check-support-for-multiple-dictionaries-os-x-and-non-en-us-locales/24937/18).

as-cii commented:
Hey @dmoonfire, thanks so much for all the efforts you've put into this: this looks great!

> Disabled the task-based handling because of passing plugins.

Could you expound a bit on this? If we want to keep using a `Task` (which seems reasonable), I wonder if we could let service authors expose a function that returns a path which, when required, exports a pure-function. This way we can easily pass those function paths into the `Task`, which in turn, is free to require them. What am I missing?

as-cii commented on lib/spell-check-manager.coffee:
```diff
@@ -0,0 +1,306 @@
+    @init()
+
+    # We need a couple packages.
+    multirange = require 'multi-integer-range'
```
Could you clarify on why we need the `multi-integer-range` library? 

Anyways, this is an interesting problem, as it would seem we'd need to define some rules for how the returned ranges need to be interpreted. For example: if a range is marked as incorrect, but there's an intersecting range that's marked as correct by another provider, what should we do? Is the whole range correct or incorrect? I guess we might define some importance rules in order to prevent this kind of mismatching.

I also wonder if it might be simpler to just expose a word-based API: spell checking, [as defined on Wikipedia],(https://en.wikipedia.org/wiki/Spell_checker) extracts words and then compares them against some kind of dictionary, so maybe we could avoid introducing this extra complexity by just having correct and incorrect words? Then, we should easily be able to identify which ranges are valid misspellings and whatnot.

Thoughts? :thought_balloon:

dmoonfire replied:
I thought about a word-based API, but then I was thinking that the overhead of native interop would be more inefficient. At least in C# and Java, there is more overhead to jump to the native libraries (as opposed to a single large string once) and I thought a word-by-word would be excessive. But, if that isn't a problem, then the `multi-integer-range` wouldn't be needed because we could stop earlier.

My general rule was if any checker says it is correct, it is correct. We do that in lines [45-47](https://github.com/dmoonfire/spell-check/blob/master/lib/spell-check-manager.coffee#L45-L47) by a simple union of indexes. Known words and `spell-check-project` both only do correctness checking. I mainly union'd those together because it was more efficient than just inverting the range and then adding them into the second rule (below).

The second rule was a word was incorrect if every checker says it is incorrect. The multiple locale checkers use this rule. To do that, we keep a integer range of incorrect words for each checker then get the intersection of all of them to identify which words are incorrect in every single one. I mainly used `multi-integer-range` to make the intersection logic a bit easier to understand using set theory instead of managing it myself (I really hate writing set operations :P).

Doing word-by-word would prevent [#13](https://github.com/atom/spell-check/issues/13) but I think that is a grammar checker thing, not a spell-checking.

as-cii replied:
Got it. I am not sure about the possible overheads of crossing the native boundary, but I'd assume that passing only reference types on the way in might be efficient enough? We might define the `check` function as:

```js
function check (words: [string]) -> [bool] {
}
```

Collecting all the `check` results enables us to have a set of words which are valid/invalid, and then we can convert them back to ranges in js land. Thoughts? :thought_balloon:

dmoonfire replied:
Now I have opinions about the word API. I banged up a quick [gist](https://gist.github.com/dmoonfire/eb9cc49a2729f78820f2) which has a couple implementations with a low-resolution timing (Date - Date). Mainly, it creates 100k words via `lorem-ipsum` and then puts it through four different combinations:

1. The entire buffer in one shot. (493 ms)
2. Each words, one at a time. (989 ms)
3. With a small `lru-cache` (1535 ms)
5. With an unlimited life cache. (45 ms)

Now, the `lru-cache` is probably implementation-specific, but it gives a rough idea of caching the results of the more common words being checked (once "a" is correct, "a" will always be right).

For the unlimited cache: Using my own novel as an example, I grabbed the first ten chapters which came out to 22k words but only 3.9k distinct words. Storing the last 5k words in memory and then wiping the cache if it exceeded it would allow you to probably easily use a fast cache and take advantage of that 45 ms lookup for the entire buffer. I'd probably say any word over 20 characters is always rechecked and skips the cache to avoid polluting it with obscure words.

The idea of the cache is that most languages have a certain redundancy (articles, prepositions, etc) that fills in most of the language.

In the worst case, checking every word on 100k words is about double what it is to send the entire buffer as a single string.

An advantage of using a word-by-word is that we could eventually have a hook that lets a grammar give us a different rule for tokenization, which would allow us to do PascalCase and camelCase (https://github.com/atom/spell-check/issues/91).

I think with a small cache (5k words, settable via config), it would be advantageous to switch to a word-by-word. It would let us drop the dependency on `multi-integer-range` and simplify the logic for marking up the errors. If you are okay with that, I'll rework the implementation and push it back up. If so, do you want it as a delta against this one or rebase it off the head as a single commit? (I'd squash everything before asking you for the final pull.)

as-cii replied:
> In the worst case, checking every word on 100k words is about double what it is to send the entire buffer as a single string.

If we manage to wrap this inside a `Task`, I feel like this might already be fine. Moreover, we can let the UI thread be updated at regular intervals, so that it doesn't look like no spell-checking is being performed in the case of a long text. :racehorse: 

Also, I am wondering if you could setup a test where, instead of crossing the native boundary **for each word**, we pass an array of strings (the words), and return an array of booleans. It seems like it could greatly improve performance because I'd assume `strings` get passed by reference, and `bool` is a really cheap data type. What do you think?

> think with a small cache (5k words, settable via config), it would be advantageous to switch to a word-by-word. It would let us drop the dependency on multi-integer-range and simplify the logic for marking up the errors.

I think it'd be nice to avoid the cache, just so we can keep the memory footprint to a minimum. Also, it seems like we would need to invalidate this cache when e.g. new packages that provide spell-checking get activated, and I'd :heart: to avoid this extra complexity if we can.

> do you want it as a delta against this one or rebase it off the head as a single commit? (I'd squash everything before asking you for the final pull.)

Personally, I think it's fine to avoid rebasing: feel free to just add new commits on top of the ones you've already pushed, so that this conversation can be read in the future within its context.

as-cii commented on lib/spell-check-manager.coffee:
```diff
@@ -0,0 +1,306 @@
+      if intersection is null
+        intersection = incorrect
+      else
+        intersection.intersect(incorrects[index])
```
Should we avoid using index here and just call `intersection.intersect(incorrect)`?

as-cii commented on lib/spell-check-manager.coffee:
```diff
@@ -0,0 +1,306 @@
+    # list. This allows us to have correct-only providers as opposed to only
+    # incorrect providers.
+    if correct.ranges.length > 0
+      intersection.subtract(correct)
```
Related to https://github.com/atom/spell-check/pull/120/files#r57576816, I guess I am not clear on what it means to subtract a range from another range? e.g.

```js
function provider1 () { 
  return {invalidRange: Range([0, 1], [0, 20])}
}

function provider2 () {
  return {validRange: Range([0,3], [0, 18])}
}
```

Should we mark `Range([0,1], [0,3])` and `Range([0, 18], [0, 20])` as incorrect?

dmoonfire replied:
The writer of `multi-integer-range` doesn't name their functions they way I expected. Their `subtract` function is really a [complement](https://en.wikipedia.org/wiki/Complement_(set_theory)) function in that all the items in set B are removed from set A.

From your example, I'm a bit confused on the input since I would expect the first to be [0, 20] and the second to be [0, 18] since there should never be an overlap in the resulting ranges.

Regardless, the resulting invalid range should only be [19,20] since the [0-3] of the correct range will completely remove the [0,1] of the invalid range. And the [0-18] will remove the [0,1] and the [18,18] of the invalid ranges leaving [19,20].

as-cii replied:
Sorry, I think I should have clarified the meaning of the `Range` syntax I used. `Range` in Atom is defined as `Range(start: Point, end: Point)`: writing `Range([0, 1], [0, 3])` means `a range that spans from {row: 0, column: 1} to {row: 0, column: 3}`.

> From your example, I'm a bit confused on the input since I would expect the first to be [0, 20] and the second to be [0, 18] since there should never be an overlap in the resulting ranges.

If the contract is that two different checkers could never return overlapping ranges, and assuming that performance-wise crossing the native boundary could be fine, what do you think about enforcing it by using that word-based API I was talking about in the comment above?

dmoonfire replied:
Nope, I read it wrong mainly because we have two types of ranges which are based on two entirely different sets. :) The `spellchecker` doesn't use `Range([startRow, startCol], [endRow, endCol])`. Instead, it [returns an array of character ranges](https://github.com/atom/spell-check/blob/master/lib/spell-check-handler.coffee#L7) `[[startIndex, endIndex], ...]`. There was logic to convert those text indexes into row/col ranges [below it](https://github.com/atom/spell-check/blob/master/lib/spell-check-handler.coffee#L9-L30).

I went with the text indexes approach to match `spellchecker` which why I went with `multi-integer-range`. Once I have a single range of incorrect words, then I [convert them to row/col ranges](https://github.com/dmoonfire/spell-check/blob/master/lib/spell-check-manager.coffee#L89-L129).

I expect a single checker so not return overlapping words but the code handles it even if it does. I do expect two checkers to include overlapping ranges which is why I do the intersection stuff.

I'm find with word-based API, it would solve this nicely, but I want to set up a quick interop test. I want to see if there is a noticeable overhead on 10k 1-word calls verses a single 10k word call. And to see if caching would make sense in this case (most languages use the same 30-40 words repeatedly).

dmoonfire commented:
@as-cii, glad you like it so far. I'll admit, my understanding of Tasks is probably the weakest part of this entire thing, so I may have missed something that makes all this possible. :)

When we allow the user to enter a list of locales into the system, I have a listener to catch those changes so I can update the plugins. It starts with the hook in [main.coffee](https://github.com/dmoonfire/spell-check/blob/master/lib/main.coffee#L18-L21) which then goes into the [manager](https://github.com/dmoonfire/spell-check/blob/master/lib/spell-check-manager.coffee#L287-L292). That clears out the `localeCheckers` in the manager which is then reloaded the first time it is needed.

Assuming the manager is in the Task, I couldn't find an example of passing configuration information or requests to reload from the main Atom process into the Task so it knows to reload all the locales before checking.

This problem also hung me up on the `KnownWordsChecker` since we pass in the list of known words into that checker (along with a flag to see if it should provide the "add to dictionary" option, etc).

I was also using the service interface to get all the additional checkers (such as `spell-check-project`, you can get that from apm) that the user loaded a package for. These are `require`d by the Atom process and passed in as objects, but I couldn't figure out how to find the path to the module from the object. I could have each checker provide its `require` path, but that felt like a code smell to me.

Related to that was also how to pass configuration changes for *those* plugins through the manager to the corresponding plugin on the other side of the Task so they know to update their configurations, reload files, etc. With the locales and known words, I know what to listen for, but arbitrary plugins (`spell-check-project` or what I was envisioning a user-specific dictionary as `spell-check-user`) wouldn't be known ahead of time.

I'd like to use Task, it would cut down on the loading time significantly, I'm just missing the required knowledge of how to make it work.

Thank you.

paulcbetts commented on lib/system-checker.coffee:
```diff
@@ -0,0 +1,91 @@
+      return
+
+    # Initialize the spell checker which can take some time.
+    @spellchecker = new spellchecker.Spellchecker
```
Awesome that you used the new API instead of the deprecated one!

dmoonfire commented:
Well, I accidentally rewrote the original commit, but I left the files as they were so the above comments remain true (line numbers and all), but sadly not the commit hash.

I believe I have implemented everything you suggested.

I did put in a short-term cache that is used only for the length of a single check on the buffer. With a 25k word chapter, it made a slightly noticeable difference that I felt it was worthwhile. We don't have to worry about invalidation because the cache is never persisted beyond the call and (ideally) the correctness of a word shouldn't change from paragraph five to paragraph fifty-five.

I updated it to handle the array-based checking, but I did not change the spellchecker module to do that. It simply checks every word as an individual call. From my earlier tests, it seemed to produce fairly reasonable performance.

I didn't touch the loading methods, only how they interacted. A lot of the code drastically changed with the inclusion of the `Task`, it added a lot more complexity in some aspects to temper the simplification from going to an array-based checking.

jeancroy commented:
There's some discussion on a word by word API in this PR comments.
I just wanted to add spiting unicode words is not easy, I expect CJK word to be even harder.

In fact spell checker used to be word by word but was changed here
https://github.com/atom/node-spellchecker/pull/27

dmoonfire commented:
@jeancroy One of the reasons I added the code I did was to handle my personal needs (grave, acute, and macrons), but I think a more holistic approach would be far better.

I don't know if using something like [XRegExp](https://github.com/slevithan/xregexp) (MIT) would be a better choice. ES6's new `/u` operator was much better at splitting things than the ES5, which would give a reason to wait until that operator was available inside Atom. It seemed like a decision that needed discussion, which is why I went with a "kinda works" approach.

There is also the problem of things like "bob's" should be one word for most dictionaries. I did an English-centric approach, but I don't like it.

Thank you.

jeancroy commented:
So this add back the word splitting on atom side instead of spellchecker side ?

With regards to `wRegexClass = '[\\w\\u0080-\\u00FF\\u0100-\\u017F\\u0180-\\u024F]'`

Other parts of atom define a list of non word character then define words character as anything not on that list. I think it make sens, I'd prefer to break the approximation on fancy ponctuation than not supporting some non latin alphabet.

dmoonfire commented:
I wasn't aware of the https://github.com/atom/node-spellchecker/pull/27 and was only going based on to feedback from the previous https://github.com/atom/spell-check/pull/120 review from @as-cii.

This does pretty much go back to tokenization on the Atom/JS side which sounds like it may be a problem. I did have a small cache involved, which would cover most natural language-based usages; repeated words are only checked once for a given buffer check which means we wouldn't include the same word repeatedly while processing the lines. For a given line, though, all the unknown words were `words.join(" ")` and passed to `node-spellcheck` which reduced the interop overhead. This approach doesn't work for opening up the `/usr/share/dict/word` because there is no redundancy and every word is on its own line, so that would be the worse-case scenario.

I guess the big question is which input should I optimize for? Line-based, word-for-word, single buffer?

With regards to `wRegexClass = '[\\w\\u0080-\\u00FF\\u0100-\\u017F\\u0180-\\u024F]'`

I wasn't entirely sure which "it" you meant, but I presume you'd like me to use the "non-word characters" approach instead of the word approach. I was mostly following how JS handled `\w` by including additional characters in the class instead of reversing the class, but I can find an example of a non-word and split on that instead. But a lot of that answer depends on the first one because if I go back to passing the entire text into the checker, it's a moot point. :)

as-cii commented:
Thanks for updating this pull-request, @dmoonfire, and thanks @jeancroy for mentioning the potential issues we might have if we switch to a word-based API.

Originally, I was mainly concerned about performance, and that's why I suggested to pass arrays as opposed to cross the native boundary for each word. However, the i18n concern feels even more important, because I underestimated the fact that a word represents a different concept based on the language, especially when it comes to non-Western locales.

Hence, it seems like we cannot simply pass words into spell-checkers, but each one should rather need to handle their own concept of word, which might take us back to the previous design. A question still remains though: how would we handle cases where two spell-checkers return inconsistent results? For instance: 

* `spell-checker-1` might return `Range({row: 1, column: 3}, {row: 1, column: 10}) = "incorrect"`.
* `spell-checker-2` might return `Range({row: 1, column: 6}, {row: 1, column: 14}) = "correct"`.

In that case there's a potential mismatch, so I guess we could either:

1. Consider the union of the reported ranges to be correct.
2. Let the spell-checker with the highest priority win.

In the above case it seems like doing 1) would be okay, because it's likely that `spell-checker-1` is using a different locale than `spell-checker-2`, so I see how always favoring correct reports over incorrect ones could make sense. In the original implementation I was concerned about using the `multi-integer-range` library and how the performance of those _intersect queries_ could have affected the speed of this package, but I believe we can actually use `marker-index` to perform the same job. The logic might roughly look like the following (pseudocode):

```js
let index = new MarkerIndex
for (let checker of this.checkers) {
  for (let result of checker.check(text)) {
    let intersectingResults = index.findIntersecting(result.range.start, result.range.end)
    for (let intersectingResultId of intersectingResults) {
      let intersectingResult = markers.get(intersectingResultId)
      if (intersectingResult.incorrect && result.correct) {
        // remove intersecting result and add current result
      } else if (intersectingResult.incorrect && result.incorrect) {
        // remove intersecting result, and add union of the two results
      } else if (intersectingResult.correct && result.correct)
        // remove intersecting result, and add union of the two results
      } else { // intersectingResult.correct && result.incorrect
        // break
      }
    }

    if (intersectingResults.length === 0) {
      // add current result
    }
}
```

All the `findIntersecting` queries should be executed in `O(log n)`, so the above routine will take `O(mn log n)`, where `m` is the number of checkers, and `n` is the number of results. We should probably perform the translation to bi-dimensional coordinates in a later pass, when we're sure that the marker index contains only _valid_ spell-check results.

@dmoonfire: thanks again for your great work on this (and sorry for revisiting my original suggestions), I really think this is going to be awesome! :sparkles:

dmoonfire commented:
@as-cii: I like talking about these things. I want to make sure we cover everything since we all bring in different experiences and skills to the table. In the end, this is all about making a good library, not personal egos. :) I also don't mind refactoring.

As for changing/revisiting opinion, that's why we discuss things. As new information is presented, we adapt.

*A question still remains though: how would we handle cases where two spell-checkers return inconsistent results?*

The general philosophy is that if one checker says its correct, it is correct. A good example for the word "danke" would be incorrect according to `en-US` but not for `de-DE`, so if you have both checkers, it will pick up `de-DE` saying it is correct and then go with that. Likewise, the known words (and project checker) only identify something as correct.

Originally, I wrote it as effectively as a tri-state (correct, incorrect, no opinion). That was to reduce the amount of data passed from the checkers. They passed a list of correct ranges, incorrect ranges, with everything else being ignored. If no checker has an opinion about a given range, then it was considered correct. This would prevent everything from being marked as incorrect if you have a bad locale.

*In the original implementation I was concerned about using the `multi-integer-range` library and how the performance of those intersect queries could have affected the speed of this package, but I believe we can actually use `marker-index` to perform the same job.*

*Question/Observation 1*

It looks like `marker-index` basically does the same thing as `multi-integer-range`, but I have a minor concern. Because we do spell-checking inside a task, everything that we pass back and forth must be JSON serializable. From the source, I saw [this line](https://github.com/atom/marker-index/blob/master/src/js/marker-index.js#L15) which implies a recursive data structure which wouldn't be JSON serializable to pass through the line. I haven't written this version (obviously), just want to know if that is going to cause a problem.

One possible fix (if `marker-index` isn't JSON serializable), is to use `marker-index`, then walk through the changes to build up the numeric ranges which are then passed across the boundary to be passed.

*Question/Observation 2*

With the proposed implementation, when `en-US` parses a 25k word document (some of my chapters are that long), everything that it says isn't incorrect (double negative) is correct. We would have to put in a correct market for those 24k words. We might be able to have a single range, but I couldn't see where `marker-index` would split a range into two when you insert a range/union it; it might be there, I don't know the library very well. Can it handle saying `Range({row: 1, column: 3}, {row: 105, column: 55})` (note the different rows) and then have `Range({row: 5, column: 10}, {row: 5, column: 15})` split it into:

* `Range({row: 1, column: 3}, {row: 5, column: 9})` = Correct
* `Range({row: 5, column: 10}, {row: 5, column: 15})` = Incorrect
* `Range({row: 5, column: 16}, {row: 105, column: 55})` = Correct

We would need the correct marking with the above implementation because of the "if any checker says correct, then its correct" to avoid `checker1` saying correct and then `checker2` marking it as incorrect. Given then, I'm concerned about memory pressure.

*Question/Observation 3*

One advantage of `multi-integer-range` is that it only works on index ranges, which is the output of the checkers. It has relatively low overhead while still allowing for set-based operations. The internal implementation is just an array of integer pairs. `marker-index`, as I read it, is a binary tree which would have a larger overhead for large documents.

Assuming that you have a 1k word document with alternative words correct:

* The `marker-index` would have 1k regions, half correct and half incorrect, in a binary tree.
* The `multi-integer-range` would have the same number of ranges, as integer pairs in two separate arrays.

For a given `marker-index` range, we would have one `Node` which has [this information](https://github.com/atom/marker-index/blob/master/src/js/node.js#L3-L12):

```
constructor (parent, leftExtent) {
    this.parent = parent
    this.left = null
    this.right = null
    this.leftExtent = leftExtent
    this.leftMarkerIds = new Set()
    this.rightMarkerIds = new Set()
    this.startMarkerIds = new Set()
    this.endMarkerIds = new Set()
    this.priority = null
    this.id = this.constructor.idCounter++
  }
```

For the each of the same ranges, `multi-integer-range` would only have `[start: number, stop: number]`.

jeancroy commented:
I like the idea of propagating "valid" state, because then union of spell checker results is sort of equivalent to a spell checker with union of dictionaries. 

However I'm not sure how I feel about propagating true state to larger area of text. For example compound word. Some part of the word may be properly spelled while another isn't. Overall the word isnt properly spelled.

Dictionary A: Detect English words, ignore hyphen
Dictionary B: Detect commonly miss-hyphenated words

--- 
Sentence: " The South-American rain forest is"

Dictionary A: { "South" : valid,  "American ": valid }
Dictionary B: { "South-American ": invalid } // Correction "South American"

Idealy valid state should not propagate.

----

Or if above is too much of an edge case...

Sentence: " South-Omerecan"

Dictionary A: { "South" : valid,  "Omerecan": invalid}
Dictionary B: { "South-Omirecan": invalid } // Correction "South American"

Idealy valid state of "South" should not propagate to  "Omerecan" using invalid state of "South-Omirecan" as a bridge.

dmoonfire commented:
I'm not sure how to implement that and still correctly handle "danke cheese" with `en-US` and `de-DE` since each one will mark opposite words as correct/incorrect.

In your examples, with the current implementation, you would have a misspelled wavy line underneath the "-". But the corrections wouldn't work since we pass the full text of the marker ("-") to the checker for suggestions.

One approach is if we used priority to determine correct/incorrect. If two checkers share a priority, then their results are merged so corrects will override, but when checker of a higher priority says it is incorrect, then it is incorrect regardless of what lower priority checkers say. That would mean there would be a lot in priority 100 (the locale checker) when you are merging multiple languages together. The only way that would work is if your second checker (the hyphenated one) *only* identified incorrect hyphenated words instead of trying to also do normal language checking. Otherwise, if that checker isn't multi-locale aware, it would incorrectly mark non-English words as incorrect even though the user had `de-DE`.

jeancroy commented:
The above problem exist because some dictionary consider part of word while
another consider compound word.

I don't think a dictionary will consider danke cheese as a compound word

On Thu, May 5, 2016, 11:24 Dylan R. E. Moonfire <notifications@github.com>
wrote:

> I'm not sure how to implement that and still correctly handle "danke
> cheese" with en-US and de-DE since each one will mark opposite words as
> correct/incorrect.
>
> In your examples, with the current implementation, you would have a
> misspelled wavy line underneath the "-". But the corrections wouldn't work
> since we pass the full text of the marker ("-") to the checker for
> suggestions.
>
> One approach is if we used priority to determine correct/incorrect. If two
> checkers share a priority, then their results are merged so corrects will
> override, but when checker of a higher priority says it is incorrect, then
> it is incorrect regardless of what lower priority checkers say. That would
> mean there would be a lot in priority 100 (the locale checker) when you are
> merging multiple languages together. The only way that would work is if
> your second checker (the hyphenated one) *only* identified incorrect
> hyphenated words instead of trying to also do normal language checking.
> Otherwise, if that checker isn't multi-locale aware, it would incorrectly
> mark non-English words as incorrect even though the user had de-DE.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/atom/spell-check/pull/120#issuecomment-217184820>
>

dmoonfire commented:
Oh, I know that "danke cheese" isn't a compound word. The reason I mentioned it is that checkers provide a text range of what is wrong. In your example, "South-Omerecan".

* Checker 1 returns `6-14` as incorrect.
* Checker 2 returns `0-14` as incorrect.

Using my current logic, it would then only mark `6-14` as incorrect because checker 1 said `0-5` is not incorrect. If I change that so checker 2 takes priority, then we would mark `0-14` as incorrect (which is the correct incorrect range for this example).

Using "danke cheese".

* Checker 1 (`en-US`) returns `0-5` as incorrect (danke is not English).
* Checker 3 (`de-DE`) returns `6-12` as incorrect (cheese is not German).

The behavior I'm working on is that if you use `en-US` and `de-DE` at the same time, "danke cheese" should show no incorrect words because they are correct for at least one checker.

The output of these checkers is the same, but there is nothing in the output that tells me that the first example (South-Omerecan) should favor the incorrect range while the second example (danke cheese) should merge the results.

I need some signal that indicates when to change the behavior. I have the ranges (as above) and priority (which is currently used to order suggestions). I could add a new property of the checkers to say "when I say incorrect, it is always incorrect" (example 1), but that wouldn't work if checker 2 also did normal language checking (e.g., all of English) because when it encountered "danke" in the second example, it would mark it as incorrect and it's signal said that its incorrect is absolute. I would consider that wrong since it breaks example 2.

I don't have a good answer because we don't have checkers that produce this conflicted results (we could make one though) and I think the solution would significantly increase the complexity of this PR.

diraimondo commented:
I'm testing your forked repository on my system. It works well for me but:
- I need to restart Atom in order to make effective a change in the language selection;
- I was not able to make work the usage of two languages; I've tried 'it-IT, en-US' without joy. Each language works alone;
- It would be great to have a general setting that keep disabled the spell checking on all tabs by default; this would allow to enable it when necessary.

dmoonfire commented:
@diraimondo I'll look into it. I started working on the requested changes from the last code review, so there is a chance that the master HEAD is unstable. But, I'll add those to the list since you shouldn't have to restart, two languages should be working.

Not entirely sure about the third, I assume you mean "turn off unless turned on." That would be a great feature, but it might be out of scope for this. It is large enough as it is and I'm worried about getting it accepted.

I'll be back working on this as soon as I get this novel published.

Thank you.

diraimondo commented:
@dmoonfire I'm speaking about the default "toggled on" status of the spell checking on all the tabs in the editor. I would keep as default "toggled off" so I can selectively toggle it on by necessity. 
At the moment I've patched your code get the desired behavior.

dmoonfire commented:
@as-cii:

I've been looking into your suggested implementation and it looks like it would be a significant amount of reworking of stuff, so I just wanted to make sure that is the direction you want to go in before I delve into it.

The first part is using `marker-index`. It looks like it should work but there are some complexities.

    // Create the empty index to represent all of the spelling areas.
    var MarkerIndex = require("marker-index");
    var index = new MarkerIndex();

The biggest is that we can't reuse the `markerId` to let us handle the three states we need to maintain:

* Unspecified: This will be correct in the end.
* Incorrect: This will mark it as incorrect unless replaced with correct.
* Correct: This is identified as correct regardless if there is an incorrect range.

    index.insert(1, {row: 2, column: 5}, {row: 2, column: 10})
    index.insert(1, {row: 3, column: 5}, {row: 3, column: 10})
    console.log(index.dump());
    //{ '1': { start: { row: 2, column: 5 }, end: { row: 3, column: 10 } } }

Multiple overlapping ranges are okay, though.

    // Handling duplicate ranges in the same area are perfectly fine.
    index.insert(2, {row: 3, column: 5}, {row: 3, column: 10})
    index.insert(3, {row: 3, column: 5}, {row: 3, column: 10})
    console.log(index.dump());
    //{ '1': { start: { row: 2, column: 5 }, end: { row: 3, column: 10 } },
    //  '2': { start: { row: 3, column: 5 }, end: { row: 3, column: 10 } },
    //  '3': { start: { row: 3, column: 5 }, end: { row: 3, column: 10 } } }

Given that, I think I would have to implement it as two separate IndexMarker (`correctIndexMarker` and `incorrectIndexMarker`) and handle the union/intersection logic independently for both. This would let me take advantage of sparse structures by not having an "undefined" IndexMarker and basically follow the same logic as I had for the `multi-integer-range`.

One reason I went with `multi-integer-range` is because it dealt with the union and intersecting logic for me instead of having me code it. That library included unit tests, which are a good thing. This also kept the code (relatively) smaller and easier to follow. IndexMarker doesn't have that, so it would increase the complexity of the file and I would have to verify my logic (more moving parts).

Another reason I original used `multi-integer-range` is that spelling works with only character offsets not the row/column references used by Atom. It just did the translation to row/column after all the merging was done, which I felt was more efficient. IndexMarker uses an internal tree to handle that where `multi-integer-range` handled it as a single array of duples.

As I mentioned in *Observation 3*, there also appears to be a significant difference in memory between the two implementations. With 10-20k word files, that could make a difference.

Thoughts? Still want me to go forward with IndexMarker? :thought_balloon:

dmoonfire commented:
@diraimondo Would you be willing to look at the latest http://github.com/dmoonfire/spell-check-test (or the one uploaded to apm (v0.77.2)? I got my code pretty much merged up and found a few events and related to working which may have impacted your problem with two languages.

If it doesn't work, can you give me the debug console's output related to spell-checking?

Thank you.

diraimondo commented:
@dmoonfire I can confirm that the spell-check-test package v.0.77.2 uploaded on apm is working well for me. Two languages work well together and a change of languages is promptly applied on the environment.

About my requirement to have toggled off the spell checking on any new tab or opened one, I understand that this could be intended out-off-scope for you work but, as you known well the structure of the plugin, it would be great if you could suggest a code snip to put in my init.coffee file to make it happen... Thanks.

as-cii commented on lib/spell-check-handler.coffee:
```diff
@@ -1,32 +1,34 @@
-SpellChecker = require 'spellchecker'
+# This is the task local handler for the manager so we can reuse the manager
+# throughout the life of the task.
+SpellCheckerManager = require './spell-check-manager.coffee'
```
Can we remove the `.coffee` extension here?

as-cii commented on lib/spell-check-manager.coffee:
```diff
@@ -0,0 +1,384 @@
+    misspellings = []
+
+    if incorrects.length is 0
+      return {id: args.id, misspellings}
```
What do you think about returning an empty array here without using the `misspellings` variable? Like:


```coffee
return {id: args.id, misspellings, []}
```

Then we can defer the variable declaration to later when we actually need it.

as-cii commented on lib/spell-check-manager.coffee:
```diff
@@ -0,0 +1,384 @@
+
+    # If we have no intersection, then nothing to report as a problem.
+    if intersection.length is 0
+      return {id: args.id, misspellings}
```
Same as https://github.com/atom/spell-check/pull/120/files#r74591382, can we return `misspellings: []` here?

as-cii commented:
Hey @dmoonfire, so after looking at the code a second time I think we might merge this. It's a pretty big pull-request and, as with all redesigns, I am somewhat concerned we might introduce some regression. The work you've done is pretty solid, though, so we might let this bake for a while on master and see if we're introducing any issue. 

With the bundled spell-checker it works great on my machine. ✨ Is there any package that provides spell-check that I might use to test this out locally? 

In the meantime could you also update this pull-request's description to match what we're currently doing in code? (i.e. the description says we're not using a `Task`, but I think we are).

Apologies for making you wait so long and thanks for the great work on this, looking forward to 🚢ping it!

dmoonfire commented:
Thank you.

Yes, if you use `spell-check-project`, it provides a project-level dictionary support. Basically the same as known words but writes out to `language.json` at the project root. It handles multiple roots fairly well too. If you include that and then you go to a misspelled word and `Control-:` it, there will be an item on the bottom to add to project dictionary.

I'll get those changes done during my lunch break.

as-cii commented:
Thanks @dmoonfire! ⚡ ⚡ ⚡

73 commented:
@as-cii Which release will that be? 1.12? (just want to have it desperately...)

as-cii commented:
Yes, this was released in Atom beta v1.12.0-beta0 and you can already try it out by downloading the latest beta from https://atom.io/beta.

73 commented:
@as-cii thanks. will try it now.

neochief commented:
It actually works! Tested in 1.12.0.beta3 and ru-RU locale.

robert-winkler commented:
Hi, spellchecking/ setting locales still troubles in Fedora25, atom 1.12.6. [#129](https://github.com/atom/spell-check/issues/129).
Any idea, what could be wrong?
Installing the spell-check-test package fails with /bin/electron : No such file or directory (but it is installed and present!)
Ok. Let's blame Fedora; I'll try a re-install of all packages related to electron/ atom node(js).

JanFrydrych commented:
Hi,
is there a faster way to change language than changing locales in spell-check package settings? Perhaps from command palette or from status bar like grammar selector?

michaeldayreads commented:
@dmoonfire - my apologies if this is not the correct place/forum to ask, but I am trying to determine if the ability to "add to dictionary" is something that should be part of the spell-check plugin, or if it requires another plug-in in addition, or if all that has been added to close #10 is the interface to make it possible for an additional plug in to add custom words to the dictionary.

If the functionality exists, I might have stumbled across a bug that prevents it from working or perhaps am experiencing user error/additional packages that an update to the README (which I would be happy to make) would fix.

I'd also be open to working to add that functionality if it is still needed and if there is anyone in the community who could mentor me in making a change to an Atom plugin.

dmoonfire commented:
The "Add to Dictionary" is actually provided by the individual plugins. For example, the system checker doesn't (we can't add words to the system dictionary) but the known words does (we control that via configuration). Also, my project-based checker also provides words.

To implement it:

1. Have the plugin return `true` for the `providesAdding` property. ([Example](https://github.com/atom/spell-check/blob/master/lib/known-words-checker.coffee#L26))
2. Implement `getAddingTargets` in the plugin. ([Example](https://github.com/atom/spell-check/blob/master/lib/known-words-checker.coffee#L39-L43), [project](https://github.com/dmoonfire/spell-check-project/blob/master/lib/project-checker.coffee#L77-L78))
3. Implement `add` in the plugin. ([Example](https://github.com/atom/spell-check/blob/master/lib/known-words-checker.coffee#L45-L48), [project](https://github.com/dmoonfire/spell-check-project/blob/master/lib/project-checker.coffee#L80-L96))

As for getting everything working, I'm willing to add my two cents. I have a pretty high workload (surprise release at work, plus publishing two books), but I usually can answer questions or give suggestions without a problem.

ethernal commented:
I am not sure if I should create a Feature Request but:

on Windows 7 the Locale Path actually works, contrary to what description says (and I'd love it to work on Windows 10 as well) 

The problem with Locale Path is that it does not accept relative paths. I have used ZIP for Atom, prepared portable only package but I have now hard-coded the path to dictionaries in the plugin settings.

Possible solutions:
* Allow relative pathis in Locale Path
* Search `.atom` folder for `dictionaries` or `dict` subfolder
* On W10 (did not test yet) if Locale Path is set it should take precedence over OS settings? Or allow us to set priority (local dictionaries/ OS dictionaries)

What do you think?
